From: Stefan Bader <stefan.bader@canonical.com>
Date: Wed, 9 Apr 2008 21:40:32 +0000 (-0400)
Subject: UBUNTU: SAUCE: Add blacklist support to fix Belkin bluetooth dongle.
X-Git-Url: http://kernel.ubuntu.com/git?p=ubuntu%2Fubuntu-intrepid.git;a=commitdiff_plain;h=0e079f51c461f48d9eceb244da7d5081a577cec3

UBUNTU: SAUCE: Add blacklist support to fix Belkin bluetooth dongle.
Bug: #140511

The Belkin bluetooth dongle unfortunately shares the vendor and device id with
the network adapter which causes lockups whenever the bluetooth dongle is
inserted. The preferred way upstream would be a mor detailed match function
but this will probably take more time. For the time being this helps people
with the bluetooth adapter.

Signed-off-by: Stefan Bader <stefan.bader@canonical.com>
---

Index: 2008-devel/drivers/net/usb/pegasus.c
===================================================================
--- 2008-devel.orig/drivers/net/usb/pegasus.c
+++ 2008-devel/drivers/net/usb/pegasus.c
@@ -1285,6 +1285,24 @@ static void check_carrier(struct work_st
 	}
 }
 
+static int pegasus_blacklisted(struct usb_device *udev)
+{
+	struct usb_device_descriptor *udd = &udev->descriptor;
+
+	/* Special quirk to keep the driver from handling the Belkin Bluetooth
+	 * dongle which happens to have the same ID.
+	 */
+	if (udd->idVendor == VENDOR_BELKIN && udd->idProduct == 0x0121) {
+		if (udd->bDeviceClass == USB_CLASS_WIRELESS_CONTROLLER) {
+			if (udd->bDeviceProtocol == 1) {
+				return 1;
+			}
+		}
+	}
+
+	return 0;
+}
+
 static int pegasus_probe(struct usb_interface *intf,
 			 const struct usb_device_id *id)
 {
@@ -1296,6 +1314,12 @@ static int pegasus_probe(struct usb_inte
 	DECLARE_MAC_BUF(mac);
 
 	usb_get_dev(dev);
+
+	if (pegasus_blacklisted(dev)) {
+			res = -ENODEV;
+			goto out;
+	}
+
 	net = alloc_etherdev(sizeof(struct pegasus));
 	if (!net) {
 		dev_err(&intf->dev, "can't allocate %s\n", "device");
